### PR TITLE
Support for nested rules

### PR DESCRIPTION
**Pull request by SerjoPepper**

### CODE REVIEW

File: test/index.js [hunk 1]
```diff
@@ -51,6 +52,33 @@ test('nested extend', function () {
...
+    '&:hover': {
+      float: 'left',
+      extend: [{ color: 'black' }]
+    },
+    '&__child': {
+      height: '10px'
+    }
+  }
+  var sheet = jss.createStyleSheet({
+    'a': {
```
Comment (kof) [inline test/index.js]: no need for quotes

File: test/index.js [hunk 2]
```diff
@@ -51,6 +52,33 @@ test('nested extend', function () {
   equal(sheet.toString(), 'a {\n  float: left;\n  display: none;\n  width: 1px;\n}')
 })
 
+test('nested extend with jss nested', function () {
+  var b = {
+    '&:hover': {
+      float: 'left',
+      extend: [{ color: 'black' }]
+    },
+    '&__child': {
```
Comment (kof) [inline test/index.js]: bad selector

File: test/index.js [hunk 3]
```diff
@@ -51,6 +52,33 @@ test('nested extend', function () {
   equal(sheet.toString(), 'a {\n  float: left;\n  display: none;\n  width: 1px;\n}')
 })
 
+test('nested extend with jss nested', function () {
```
Comment (kof) [inline test/index.js]: why overcomplicate the test? In this particular one you need to test the __merge__ from 2 extends. A lot of unrelated stuff here.

File: src/index.js [hunk 4]
```diff
@@ -1,5 +1,7 @@
-const warn = console.warn.bind(console) // eslint-disable-line no-console
+import merge from 'lodash.merge'
```
Comment (kof) [inline src/index.js]: lodash.merge is too big and handles way more cases than we need, I would reimplement it here with something minimal

**Comment by SerjoPepper** [comment]
Update review with fixes

File: src/index.js [hunk 5]
```diff
@@ -25,22 +25,24 @@ function extend(rule, newStyle, style) {
       if (prop === 'extend') {
         extend(rule, newStyle, style.extend.extend)
       }
-      else if (isObject(style[prop]) && isObject(style.extend[prop])) {
-        newStyle[prop] = merge(
-          extend(rule, {}, style[prop]),
-          extend(rule, {}, style.extend[prop])
-        )
+      else if (isPlainObject(style.extend[prop])) {
```
Comment (kof) [inline src/index.js]: we don't expect non-plain objects here ... a simple typeof should do ...

File: test/index.js [hunk 6]
```diff
@@ -51,6 +53,64 @@ test('nested extend', function () {
   equal(sheet.toString(), 'a {\n  float: left;\n  display: none;\n  width: 1px;\n}')
 })
 
+test('nested extend', function () {
```
Comment (kof) [inline test/index.js]: still lots of unneeded props in the objects ...

Comment (kof) [inline test/index.js]: ah no its fine, you are also checking if the override of width worked.

File: test/index.js [hunk 7]
```diff
@@ -51,6 +53,64 @@ test('nested extend', function () {
...
+  }
+  var d = {
+    extend: c,
+    '&:hover': {width: '2px'}
+  }
+  var sheet = jss.createStyleSheet({
+    a: {
+      extend: d,
+      width: '2px',
+      border: {
```
Comment (kof) [inline test/index.js]: cool

File: src/index.js (hunk 5 above)
Comment (SerjoPepper) [inline src/index.js]: how about unexpected "null" objects and arrays from some plugins (jss-expand)? I don't think we want to merge this `padding: [20, 10]`

Comment (kof) [inline src/index.js]: oh yes, you are right. Considered the cases we have, I did benchmarks: https://github.com/cssinjs/jss/blob/master/benchmark/tests/is-object.js

This implementation is 6.5 times faster than yours:

```
  function isObject(obj) {
    return obj && typeof obj === 'object' && !Array.isArray(obj)
  }
```

Lets use it.

**Comment by kof** [comment]
merged

**Comment by kof** [comment]
I have just found a bug, which has been introduced by this pr.

```javascript
test('ensure override order', function () {
  var a = {
    float: 'left',
    color: 'red'
  }
  var sheet = jss.createStyleSheet({
    a: {
      extend: a,
      float: 'right'
    }
  }, {named: false})
  ok(sheet.getRule('a'))
  equal(sheet.toString(), 'a {\n  float: right;\n  color: red;\n}')
})
```

Rules own properties should always overwrite the extended.

**Comment by kof** [comment]
We even have wrong tests, which show that the extend can override the rule.

**Comment by kof** [comment]
Creating a fix in a bit.

**Comment by SerjoPepper** [comment]
ok, I can make pr. i did not know about this behavior, also I thought that extend have priority over own props

**Comment by kof** [comment]
already fixed and published

**Comment by kof** [comment]
Which means that its a documentation lack.
